src/common/utxobased/engine/makeUtxoEngineState.ts

```diff
@@ -240,27 +307,256 @@ const setLookAhead = async (args: SetLookAheadArgs): Promise<void> => {
+      deferred
+    )
+  }
+
+  // Loop to process addresses to utxos
+  for (const [address, state] of utxosCache) {
+    // Check if we need to fetch address UTXOs
+    if (!state.fetching) {
+      state.fetching = true
+
+      utxosCache.delete(address)
+
+      // Fetch and process address UTXOs
+      const wsTask = await processAddressUtxos({
+        ...args,
+        ...state,
+        address
+      })
+      return wsTask
+    }
+  }
+
+  // Check if there are any addresses pending to be subscribed
+  if (addressSubscribeCache.size > 0 && !taskCache.addressWatching) {
+    // Loop each address that needs to be subscribed
+    for (const [address, path] of addressSubscribeCache) {
+      // Add address in the cache to the set of addresses to watch
+      addressesToWatch.set(address, path)
+
+      if (path != null) {
+        // Add the newly watched addresses to the UTXO cache
+        utxosCache.set(address, {
+          fetching: false,
+          path
+        })
+        await addToTransactionCache(
+          args,
+          address,
+          path.format,
+          path.branch,
+          transactionsCache
+        )
+      }
+    }
+
+    taskCache.addressWatching = true
+
+    const deferred = new Deferred<unknown>()
+    deferred.promise
+      .then(() => {
+        addressSubscribeCache.clear()
+      })
+      .catch(() => {
+        taskCache.addressWatching = false
+      })
+    blockBook.watchAddresses(
+      Array.from(addressesToWatch.keys()),
+      (response: INewTransactionResponse) => {
+        const path = addressesToWatch.get(response.address)
+        if (path != null) {
+          utxosCache.set(response.address, {
+            fetching: false,
+            path
+          })
+          addToTransactionCache(
+            args,
+            response.address,
+            path.format,
+            path.branch,
+            transactionsCache
+          ).catch(() => {
+            throw new Error('failed to add to transaction cache')
+          })
+        }
+      },
+      deferred
+    )
+    return
+  }
+
+  // filled when transactions potentially changed (e.g. through new block notification)
+  for (const [txId, state] of updateTransactionsCache) {
+    if (!state.fetching) {
+      state.fetching = true
+      updateTransactionsCache.delete(txId)
+      return updateTransactions({ ...args, txId })
+    }
+  }
+
+  // Loop unparsed utxos, some require a network call to get the full tx data
+  for (const [utxo, state] of rawUtxosCache) {
+    if (!state.fetching) {
+      state.fetching = true
+      rawUtxosCache.delete(utxo)
+      const wsTask = await processRawUtxo({
+        ...args,
+        ...state,
+        ...state.path,
+        address: state.address,
+        utxo,
+        id: `${utxo.txid}_${utxo.vout}`
+      })
+      return wsTask
+    }
+  }
+
+  // Loop processed utxos, these are just database ops, triggers setLookAhead
+  for (const [address, state] of processedUtxosCache) {
```

Comment:
Move towards the top, try to get the user to be able to spend money asap, then fetch address, then tx history.